app/src/main/java/org/asteroidos/sync/services/SynchronizationService.java

```diff
@@ -31,198 +33,241 @@
+    }
 
-                replyTo.send(Message.obtain(null, MSG_SET_STATUS, mState, 0));
-            } catch (RemoteException | NullPointerException ignored) {}
+    @Override
+    public final ConnectionState getConnectionState() {
+        ConnectionState state = null;
+        switch (mState) {
+            case (STATUS_CONNECTED):
+                state = ConnectionState.STATUS_CONNECTED;
+                break;
+            case (STATUS_CONNECTING):
+                state = ConnectionState.STATUS_CONNECTING;
+                break;
+            case (STATUS_DISCONNECTED):
+                state = ConnectionState.STATUS_DISCONNECTED;
+                break;
+        }
+        return state;
+    }
+
+    @Override
+    public final void send(UUID characteristic, byte[] data, IConnectivityService service) {
+        mBleMngr.send(characteristic, data);
+        Log.d(TAG, characteristic.toString() + " " + Arrays.toString(data));
+    }
 
-            editor.putString(MainActivity.PREFS_DEFAULT_LOC_NAME, name);
+    @Override
+    public final void registerBleService(IConnectivityService service) {
+        boolean success = bleServices.add(service);
+        Log.d(TAG, "BLE Service registered: " + success + service.getServiceUUID());
+    }
+
+    @Override
+    public final void unregisterBleService(UUID serviceUUID) {
+        for (IConnectivityService service : bleServices) {
+            if (service.getServiceUUID().equals(serviceUUID)) {
+                bleServices.remove(service);
+                Log.d(TAG, "BLE Service unregistered: " + service.getServiceUUID());
+            }
         }
-        editor.apply();
     }
 
-    void handleUpdate() {
-        if(mDevice != null) {
-            try {
-                Message answer = Message.obtain(null, MSG_SET_LOCAL_NAME);
-                answer.obj = mDevice.getName_normalized();
-                replyTo.send(answer);
+    @Override
+    public final void registerCallback(UUID characteristicUUID, IServiceCallback callback) {
+        mBleMngr.recvCallbacks.putIfAbsent(characteristicUUID, callback);
+    }
 
-                replyTo.send(Message.obtain(null, MSG_SET_STATUS, mState, 0));
+    @Override
+    public final void unregisterCallback(UUID characteristicUUID) {
+        mBleMngr.recvCallbacks.remove(characteristicUUID);
+    }
 
-                mDevice.read(AsteroidUUIDS.BATTERY_UUID, new BleDevice.ReadWriteListener()
-                {
-                    @Override public void onEvent(ReadWriteEvent result)
-                    {
-                        if(result.wasSuccess())
-                            try {
-                                replyTo.send(Message.obtain(null, MSG_SET_BATTERY_PERCENTAGE, result.data()[0], 0));
-                            } catch (RemoteException | NullPointerException ignored) {}
+    @Override
+    public final IConnectivityService getServiceByUUID(UUID uuid) {
+        for (IConnectivityService service : bleServices) {
```

Comment:
I think that bleServices is another example of datastructure that you use for searching by key. If you use a list the lookup will be O(n) and require custom code, if you use a hashmap indexed by the key you are interested in, the lookup is a standard library call and O(1)

Reply:
Note that if you not only lookup by serviceuuid but also by each characteristic uuid you could still had multiple keys mapping to the same object in the HashMap